src/rust/server/segcache/tests/integration.rs

```diff
@@ -0,0 +1,166 @@
+    );
+    test(
+        "add value (key: 1)",
+        &[("add 1 0 0 1\r\n2\r\n", Some("STORED\r\n"))],
+    );
+    test(
+        "get value (key: 0)",
+        &[("get 0\r\n", Some("VALUE 0 0 1\r\n1\r\nEND\r\n"))],
+    );
+    test(
+        "get value (key: 1)",
+        &[("get 1\r\n", Some("VALUE 1 0 1\r\n2\r\nEND\r\n"))],
+    );
+    test(
+        "replace value (key: 0)",
+        &[("replace 0 0 0 1\r\n2\r\n", Some("STORED\r\n"))],
+    );
+    test(
+        "replace value (key: 2)",
+        &[("replace 2 0 0 1\r\n2\r\n", Some("NOT_STORED\r\n"))],
+    );
+    test(
+        "get value (key: 0)",
+        &[("get 0\r\n", Some("VALUE 0 0 1\r\n2\r\nEND\r\n"))],
+    );
+    test("get value (key: 2)", &[("get 2\r\n", Some("END\r\n"))]);
+
+    // test storing and retrieving flags
+    test(
+        "set value (key: 3)",
+        &[("set 3 42 0 1\r\n1\r\n", Some("STORED\r\n"))],
+    );
+    test(
+        "get value (key: 3)",
+        &[("get 3\r\n", Some("VALUE 3 42 1\r\n1\r\nEND\r\n"))],
+    );
+
+    // shutdown server and join
+    debug!("shutdown");
+    let _ = server.shutdown();
+}
+
+// opens a new connection, operating on request + response pairs from the
+// provided data.
+fn test(name: &str, data: &[(&str, Option<&str>)]) {
+    info!("testing: {}", name);
+    debug!("connecting to server");
+    let mut stream = TcpStream::connect("127.0.0.1:12321").expect("failed to connect");
+    stream
+        .set_read_timeout(Some(Duration::from_millis(250)))
+        .expect("failed to set read timeout");
+    stream
+        .set_write_timeout(Some(Duration::from_millis(250)))
+        .expect("failed to set write timeout");
+
+    debug!("sending request");
+    for (request, response) in data {
+        match stream.write(request.as_bytes()) {
+            Ok(bytes) => {
+                if bytes == request.len() {
+                    debug!("full request sent");
+                } else {
+                    error!("incomplete write");
+                    fatal!("status: failed\n");
+                }
+            }
+            Err(_) => {
+                error!("error sending request");
+                fatal!("status: failed\n");
+            }
+        }
+
+        std::thread::sleep(Duration::from_millis(10));
+        let mut buf = vec![0; 4096];
+
+        if let Some(response) = response {
+            if stream.read(&mut buf).is_err() {
+                fatal!("error reading response");
+            } else if response.as_bytes() != &buf[0..response.len()] {
+                error!("expected: {:?}", response.as_bytes());
+                error!("received: {:?}", &buf[0..response.len()]);
+                fatal!("status: failed\n");
+            } else {
+                debug!("correct response");
+            }
+            assert_eq!(response.as_bytes(), &buf[0..response.len()]);
+        } else if let Err(e) = stream.read(&mut buf) {
+            if e.kind() == std::io::ErrorKind::WouldBlock {
```

Comment:
I'm surprised that `ErrorKind` doesn't have an explicit equivalent for `EPOLLHUP`, why do you think that is?

Reply:
The `std::io::Read` (https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read) maps to `read()` (https://man7.org/linux/man-pages/man2/read.2.html) - so count of 0 bytes returned for EOF.

Fun fact - `mio` used to have a unix specific readiness type for HUP - but it was removed in favor of keeping the readiness/interests platform agnostic. 

In this specific call site, we're not even using epoll, the integration test just uses a plain `TcpStream` in non-blocking mode as a client.

Leaving this open as a good reminder to check my error handling in the actual server code.

Reply:
Fixed the handling for `EINTR` in the event loop.